staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1091 @@
+	attInSpecificVNSelctor := attVNIFieldName + equal + fmt.Sprint(vni)
+
+	// Build and return a selector which is a logical AND between all the selectors defined above.
+	allSelectors := []string{remoteAttSelector,
+		hostIPIsNotLocalSelector,
+		attWithAnIPSelector,
+		attWithHostIPSelector,
+		attWithIfcSelector,
+		attInSpecificVNSelctor}
+	return strings.Join(allSelectors, ",")
+}
+
+func fromFieldsSelectorToTweakListOptionsFunc(customFieldSelector string) kosinternalifcs.TweakListOptionsFunc {
+	return func(options *k8smetav1.ListOptions) {
+		optionsFieldSelector := options.FieldSelector
+		allSelectors := make([]string, 0, 2)
+		if strings.Trim(optionsFieldSelector, " ") != "" {
+			allSelectors = append(allSelectors, optionsFieldSelector)
+		}
+		allSelectors = append(allSelectors, customFieldSelector)
+		options.FieldSelector = strings.Join(allSelectors, ",")
+	}
+}
+
+func v1a1AttsCustomInformerAndLister(kcs *kosclientset.Clientset,
+	resyncPeriod time.Duration,
+	tweakListOptionsFunc kosinternalifcs.TweakListOptionsFunc) (k8scache.SharedIndexInformer, koslisterv1a1.NetworkAttachmentLister) {
+
+	attv1a1Informer := createAttsv1a1Informer(kcs,
+		resyncPeriod,
+		k8smetav1.NamespaceAll,
+		tweakListOptionsFunc)
+	return attv1a1Informer.Informer(), attv1a1Informer.Lister()
+}
+
+func v1a1AttsCustomNamespaceInformerAndLister(kcs *kosclientset.Clientset,
+	resyncPeriod time.Duration,
+	namespace string,
+	tweakListOptionsFunc kosinternalifcs.TweakListOptionsFunc) (k8scache.SharedIndexInformer, koslisterv1a1.NetworkAttachmentNamespaceLister) {
+
+	attv1a1Informer := createAttsv1a1Informer(kcs,
+		resyncPeriod,
+		namespace,
+		tweakListOptionsFunc)
+	return attv1a1Informer.Informer(), attv1a1Informer.Lister().NetworkAttachments(namespace)
+}
+
+func createAttsv1a1Informer(kcs *kosclientset.Clientset,
+	resyncPeriod time.Duration,
+	namespace string,
+	tweakListOptionsFunc kosinternalifcs.TweakListOptionsFunc) kosinformersv1a1.NetworkAttachmentInformer {
+
+	localAttsInformerFactory := kosinformers.NewFilteredSharedInformerFactory(kcs,
+		resyncPeriod,
+		namespace,
+		tweakListOptionsFunc)
+	netv1a1Ifc := localAttsInformerFactory.Network().V1alpha1()
+	return netv1a1Ifc.NetworkAttachments()
+}
+
+// attachmentMACAddr is an Index function that computes the MAC address of a NetworkAttachment.
+// Used to map pre-existing interfaces with attachments at start up.
+func attachmentMACAddr(obj interface{}) (subnets []string, err error) {
+	att := obj.(*netv1a1.NetworkAttachment)
+	return []string{generateMACAddr(att.Spec.VNI, gonet.ParseIP(att.Status.IPv4)).String()}, nil
+}
+
+func generateIfcName(macAddr gonet.HardwareAddr) string {
+	return "kos" + strings.Replace(macAddr.String(), ":", "", -1)
+}
+
+func generateMACAddr(vni uint32, guestIPv4 gonet.IP) gonet.HardwareAddr {
+	guestIPBytes := guestIPv4.To4()
+	mac := make([]byte, 6, 6)
+	mac[5] = byte(vni >> 0)
+	mac[4] = byte(vni >> 8)
+	mac[3] = byte(vni>>16) | (guestIPBytes[3] << 5)
```

Comment:
We could do even less bit twiddling if we copied `guestIPBytes[3]` directly into one of the `mac` bytes, leaving only one `mac` byte computed by bit twiddling.

Reply:
Fixed in 5aca980.